### PR TITLE
DOCK-2228 Added apptools in RSS feed

### PR DESCRIPTION
**Pull request by hyunnaye**
**Description**
RSS feed is currently missing apptools. This PR adds apptools into the RSS feed.

For the test `testAppToolRSSFeedAndSiteMap`, it tests that that the DAO creates paths correctly and apptools are loaded onto the RSS Feed and sitemap in `MetadataResource`.

**Review Instructions**
Verify that the queries to retrieve app tool paths are correct. Then confirm that the tests correctly tests apptools  being added to the RSS feed and sitemap in `MetadataResource.java`. 

**Issue**
https://github.com/dockstore/dockstore/issues/5086

Please make sure that you've checked the following before submitting your pull request. Thanks!

- [x] Check that you pass the basic style checks and unit tests by running `mvn clean install`
- [x] Ensure that the PR targets the correct branch. Check the milestone or fix version of the ticket.
- [x] Follow the existing JPA patterns for queries, using named parameters, to avoid SQL injection
- [x] If you are changing dependencies, check the Snyk status check or the dashboard to ensure you are not introducing new high/critical vulnerabilities
- [x] Assume that inputs to the API can be malicious, and sanitize and/or check for Denial of Service type values, e.g., massive sizes
- [x] Do not serve user-uploaded binary images through the Dockstore API
- [x] Ensure that endpoints that only allow privileged access enforce that with the `@RolesAllowed` annotation
- [x] Do not create cookies, although this may change in the future
- [x] If this PR is for a user-facing feature, create and link a documentation ticket for this feature (usually in the same milestone as the linked issue). Style points if you create a documentation PR directly and link that instead.

### CODE REVIEW

**Comment by denis-yuen** [comment]
> Heads-up: LGTM.com's PR analysis will be disabled on the 5th of December, 

Hmmm, that's interesting

File: dockstore-webservice/src/main/java/io/dockstore/webservice/resources/MetadataResource.java [hunk 1]
```diff
@@ -178,6 +183,7 @@ public SortedSet<String> getSitemap() {
         SortedSet<String> urls = new TreeSet<>();
         urls.addAll(getToolPaths());
         urls.addAll(getBioWorkflowPaths());
+        urls.addAll(getAppToolPaths());
```
Comment (denis-yuen) [inline dockstore-webservice/src/main/java/io/dockstore/webservice/resources/MetadataResource.java:186]: Thanks for noticing and handling the sitemap at the same time!
Could use a test as well

**Review by denis-yuen — COMMENTED** [review]
Look for `testRSSPlusSiteMap` for a test that tests the endpoints in addition to the DAO.
Bonus points if you write a version of the test that uses the openapi client rather than the swagger client since we've slowly transitioning off swagger

File: dockstore-webservice/src/main/java/io/dockstore/webservice/helpers/MetadataResourceHelper.java [hunk 2]
```diff
@@ -24,6 +41,8 @@ public static String createWorkflowURL(Workflow workflow) {
             return baseUrl + "/workflows/" + workflow.getWorkflowPath();
         } else if (workflow instanceof Service) {
             return baseUrl + "/services/" + workflow.getWorkflowPath();
+        } else if (workflow instanceof AppTool) {
+            return baseUrl + "/apptools/" + workflow.getWorkflowPath();
```
Comment (coverbeck) [inline dockstore-webservice/src/main/java/io/dockstore/webservice/helpers/MetadataResourceHelper.java:45]: I don't think this is right; I believe it should be `"/containers/"` (but please verify). An example url for an app tool on qa is `https://qa.dockstore.org/containers/github.com/common-workflow-library/bio-cwl-tools/bandage-info:release?tab=info`, with `containers`, so I think it means it has to be.

**Review by denis-yuen — APPROVED** [review]

**Review by coverbeck — APPROVED** [review]

**Comment by svonworl** [comment]
Regarding an amplification of some preexisting behavior...
Now that `getRss` combines three lists of of max length `io.dockstore.webservice.resources.MetadataResource.RSS_ENTRY_LIMIT`, it appears that the RSS feed can contain three times more entries than the limit.  Should we truncate to the limit?

**Comment by denis-yuen** [comment]
> Regarding an amplification of some preexisting behavior... Now that `getRss` combines three lists of of max length `io.dockstore.webservice.resources.MetadataResource.RSS_ENTRY_LIMIT`, it appears that the RSS feed can contain three times more entries than the limit. Should we truncate to the limit?

Looks like the limit was introduced in https://github.com/dockstore/dockstore/pull/2984#issue-511353564 due to performance concerns. Out of curiosity, prod (which doesn't have app tools) gives 

```
$ curl -o /dev/null -s -w 'Total: %{time_total}s\n'  https://dockstore.org/api/metadata/rss
Total: 0.314085s
```

File: dockstore-webservice/src/main/java/io/dockstore/webservice/resources/MetadataResource.java [hunk 3]
```diff
@@ -264,7 +277,13 @@ private String getRSS() {
         List<RSSEntry> entries = new ArrayList<>();
         for (Entry<?, ?> dbEntry : dbEntries) {
             RSSEntry entry = new RSSEntry();
-            if (dbEntry instanceof Workflow) {
+            if (dbEntry instanceof AppTool) {
```
[the remark this answers is not in the document]
Comment (hyunnaye) [reply dockstore-webservice/src/main/java/io/dockstore/webservice/resources/MetadataResource.java]: Changed!

**Comment by svonworl** [comment]
> > Regarding an amplification of some preexisting behavior... Now that `getRss` combines three lists of of max length `io.dockstore.webservice.resources.MetadataResource.RSS_ENTRY_LIMIT`, it appears that the RSS feed can contain three times more entries than the limit. Should we truncate to the limit?
> 
> Looks like the limit was introduced in [#2984 (comment)](https://github.com/dockstore/dockstore/pull/2984#issue-511353564) due to performance concerns. Out of curiosity, prod (which doesn't have app tools) gives
> 
> ```
> $ curl -o /dev/null -s -w 'Total: %{time_total}s\n'  https://dockstore.org/api/metadata/rss
> Total: 0.314085s
> ```



> > Regarding an amplification of some preexisting behavior... Now that `getRss` combines three lists of of max length `io.dockstore.webservice.resources.MetadataResource.RSS_ENTRY_LIMIT`, it appears that the RSS feed can contain three times more entries than the limit. Should we truncate to the limit?
> 
> Looks like the limit was introduced in [#2984 (comment)](https://github.com/dockstore/dockstore/pull/2984#issue-511353564) due to performance concerns. Out of curiosity, prod (which doesn't have app tools) gives
> 
> ```
> $ curl -o /dev/null -s -w 'Total: %{time_total}s\n'  https://dockstore.org/api/metadata/rss
> Total: 0.314085s
> ```

Cool.  Another reason to limit would be to cap the size of the generated RSS, but prod RSS is currently only 45kB, and our audience (and potential number of RSS subscribers) is limited, so it'll probably never be a concern.

**Review by svonworl — APPROVED** [review]

**Review by kathy-t — APPROVED** [review]

**Pull request merged**